### DISCUSSION TITLE
Localized keyboard (Czech) is missing mapping for <Ctrl-ú> (ESC control code) on macOS

### DISCUSSION THREAD

**Discussion by risa2000** [vscode-neovim/vscode-neovim#1875] [Q&A] [answered]
This is rather a question than a bug report, but I have no idea where to start.

I have recently added `vscode-neovim` to VSCode on my Windows 10 machine and was rather satisfied with the results, which inspired me to try it in VSCode on macOS (Sonoma on Apple silicon).
The important fact is that in both systems I am using Czech localized keyboard layout (on both systems the one called "Czech"). While on Windows the localized keyboard works fine, on macOS as soon as I switch to the `CZ` layout I cannot emit `ESC` control code (to quit insert mode, etc.) by using the `<C-ú>` combo.

Looking at the keyboard mapping in VSCode while using `CZ` layout shows that while in Windows the `<C-[>` is correctly remapped the `<C-ú>`,  (character `ú` in `CZ` layout is written by the same key as `[` in `US` layout), on macOS, the mapping simply vanishes and reappears when switching back to `US` layout.

Both machines run the same VSCode version 1.87.2
Few screenshot to illustrate what I mean by "mapping simply vanishes".

**Windows using `US` layout, with correct mapping:**
![neovim_escape_mapping_win10_us_kb](https://github.com/vscode-neovim/vscode-neovim/assets/1336835/394feb59-f4a7-4aab-9bb2-8b7a816b6ac8)

**Windows using `CZ` layout, with correct mapping:**
![neovim_escape_mapping_win10_cz_kb](https://github.com/vscode-neovim/vscode-neovim/assets/1336835/5ee7f8ed-e384-4ccd-83bb-6847dedfbc68)

**macOS using `US` layout and having correct mapping:**
<img width="800" alt="neovim_escape_mapping_macos_us_kb" src="https://github.com/vscode-neovim/vscode-neovim/assets/1336835/b779bcb5-7c27-4f23-b73d-b02ee2c26ccb">

**macOS using `CZ` layout and missing the `<C-ú>` mapping:**
<img width="800" alt="neovim_escape_mapping_macos_cz_kb" src="https://github.com/vscode-neovim/vscode-neovim/assets/1336835/8dbd2915-4fcb-412b-855d-a3b0a9359e1c">

What I found interesting is that the displayed mapping in VSCode changes on the fly when switching the keyboard layouts. So it looks like VSCode knows how to process the "input source" change in both systems. Since it also puts back the correct one in the `US` version, it looks like the definitions are correct, but for some reason the localized one on macOS does not make it to the actual interpretation (of the mappings) and does not work as well.

**Comment by risa2000** [comment]
The issue is still present in the latest version 1.18.14. I suspect that the culprit is not the extension, but I would appreciate any insight from the devs on mechanics involved.

I checked the keyboard shortcuts defined in `package.json` and there is only `ctrl+[` defined:
```json
{
	"command": "vscode-neovim.escape",
	"key": "ctrl+[",
	"when": "editorTextFocus && neovim.init && editorLangId not in neovim.editorLangIdExclusions"
},
```
yet, by some magic, this is exposed in VSCode in Windows with Czech keyboard layout as this:
![neovim_escape_mapping_win10_cz_kb_crop](https://github.com/user-attachments/assets/2a799594-e6c9-4e77-8211-fd51c6d2dc01)
i.e. the mapping defines `Ctrl+ú` as the escape key, which is correct. (This is physically the same key, it just writes a different character in localized keyboard layout.)

On macOS, when switching to the Czech keyboard layout the mapping simply disappears (as shown in my original post).

The questions is which part is responsible for the key definition change. It is clearly not in `package.json`, but is dynamically redefined according to the keyboard layout (the mappings in vscode window - shown on the screenshot - change immediately when switching the keyboard layout).
 1) Is neovim extension somehow inquiring the system API to translate the mapping definitions (i.e. in this case to translate `ctrl+[` to `ctrl+ú`)?
 2) Is neovim extension "hinting" VSCode with `ctrl+[` key, and VSCode remaps it to `ctrl+ú`?
 3) Where/when exactly the correct mapping for localized keyboard emerges?

The plain `neovim` (as well as `macvim`) work fine on macOS and respect the localized keyboard layout, so the issue here is related to either VSCode or neovim extension (or both).

**Reply by justinmk** [reply]
> 1\. Is neovim extension somehow inquiring the system API to translate the mapping definitions

No. It just sends the raw input to `nvim`, which then interprets it as usual. Of course, the "raw input" received by vscode-neovim is subject to whatever mappings are active in vscode itself.

> 2\. Is neovim extension "hinting" VSCode with `ctrl+[` key, and VSCode remaps it to `ctrl+ú`?

vscode-neovim sets some default mappings for `ctrl-[` but AFAIR it doesn't do anything beyond that. You can always delete these default mappings in vscode settings.


> The plain `neovim` (as well as `macvim`) work fine on macOS and respect the localized keyboard layout, so the issue here is related to either VSCode or neovim extension (or both).

You can try sending raw input to nvim by calling `:call nvim_input('<c-[>')` from vscode-neovim's `:` prompt.

**Reply by risa2000** [reply]
> > 1. Is neovim extension somehow inquiring the system API to translate the mapping definitions
> 
> No. It just sends the raw input to `nvim`, which then interprets it as usual. Of course, the "raw input" received by vscode-neovim is subject to whatever mappings are active in vscode itself.

The keyboard mapping (shown in keyboard shortcuts in VSCode on my screenshot) defines a mapping between a key input and some vscode-neovim command in some particular context (e.g. `^C` -> `vscode-neovim.escape` when `"editorTextFocus && neovim.init && editorLangId not in neovim.editorLangIdExclusions && neovim.mode == normal && neovim.ctrlKeysNormal.c && !markersNavigationVisible && !parameterHintsVisible && !inReferenceSearchEditor && !referenceSearchVisible && !dirtyDiffVisible && !notebookCellFocused && !findWidgetVisible && !notificationCenterVisible"`).

Is this mapping applied by VSCode already **before** sending "raw input" to vscode-neovim, or is it something vscode-neovim applies before passing it further to nvim? If the former then what is vscode-neovim sending to nvim?
In other words, when I press `^C` does vscode-neovim extension receive `^C` or `vscode-neovim.escape`?

Is there a way to see the raw input vscode-neovim receives from VSCode?

> > 2. Is neovim extension "hinting" VSCode with `ctrl+[` key, and VSCode remaps it to `ctrl+ú`?
> 
> vscode-neovim sets some default mappings for `ctrl-[` but AFAIR it doesn't do anything beyond that. You can always delete these default mappings in vscode settings.

I guess these might be those defined in vscode-neovim `package.json` file. There is a mapping defined for `ctrl+[`. If this is all vscode-neovim does, then changing it to `ctrl+ú` for Czech keyboard layout must be VSCode witchery.  

> > The plain `neovim` (as well as `macvim`) work fine on macOS and respect the localized keyboard layout, so the issue here is related to either VSCode or neovim extension (or both).
> 
> You can try sending raw input to nvim by calling `:call nvim_input('<c-[>')` from vscode-neovim's `:` prompt.

It is difficult to tell what exactly happens, because when I am on prompt sending "escape" does nothing. But as I wrote, nvim works fine when it receives `<c-[>` (when run in `VimR`, which is macOS "GUI" version of nvim), so I would expect it should work.

**Reply by justinmk** [reply]
> Is there a way to see the raw input vscode-neovim receives from VSCode?

enable "debug" level logging https://github.com/vscode-neovim/vscode-neovim/?tab=readme-ov-file#troubleshooting

not sure if the raw keys will be logged though. you might need to add a logging statement and then run the project. it's quite easy to setup the repo, just `npm install` and then hit F5 in vscode...

**Reply by risa2000** [reply]
> > Is there a way to see the raw input vscode-neovim receives from VSCode?
> 
> enable "debug" level logging https://github.com/vscode-neovim/vscode-neovim/?tab=readme-ov-file#troubleshooting

If I enable `trace` level in the output, I can see following sequence when pressing `i` followed by `ctrl+[` when using US keyboard layout.
```log
2025-01-05 19:58:20.187 [debug] ModeManager: Changing mode to i
2025-01-05 19:58:20.187 [debug] ModeManager: Setting mode context to insert
2025-01-05 19:58:20.187 [debug] TypingManager: Waiting for cursor completion operation before disposing type handler
2025-01-05 19:58:27.860 [debug] TypingManager: Send for: <Esc>
2025-01-05 19:58:27.861 [debug] TypingManager: Syncing buffers with neovim (<Esc>)
2025-01-05 19:58:27.862 [debug] TypingManager: Pending keys sent with <Esc>: 
2025-01-05 19:58:27.870 [debug] ModeManager: Changing mode to n
2025-01-05 19:58:27.870 [debug] ModeManager: Setting mode context to normal
```
It is not clear what is received or sent from the extension. Does
```
2025-01-05 19:58:27.860 [debug] TypingManager: Send for: <Esc>
```
mean that vscode-neovim received `<Esc>` code, or that it sent `<Esc>` code to nvim?

When doing the same with Czech keyboard, I see only this:
```log
2025-01-05 20:00:23.724 [debug] ModeManager: Changing mode to i
2025-01-05 20:00:23.724 [debug] ModeManager: Setting mode context to insert
2025-01-05 20:00:23.724 [debug] TypingManager: Waiting for cursor completion operation before disposing type handler
```
I.e. it seems no key is registered for `ctrl+[` at all.

Does this mean VSCode does not pass the keypress to the extension?

> not sure if the raw keys will be logged though. you might need to add a logging statement and then run the project. it's quite easy to setup the repo, just `npm install` and then hit F5 in vscode...

I guess the raw keys are not logged then. I do not have Node.js installed and would prefer to have it this way. Is there a way to get the raw keys without it?

**Comment by risa2000** [comment, marked as answer]
Coming back to this topic. It seems that the difference in behavior comes from the VSCode itself.

Once I figured out how to enable keyboard debugging in VSCode (`Developer: Toggle Keyboard Shortcuts Troubleshooting` command), I found out that while VSCode on Windows is able to correctly "localize" the keypress to `ctrl+[`, which is the keybinding defined in `vscode-neovim` keybindings (`package.json`), on macOS the resolution does make it only to the point of `ctrl+[BracketLeft]`, which is not defined in this extension's keybindings.

[I have asked on VSCode github, why VSCode behaves differently on the two OSes](https://github.com/microsoft/vscode/issues/237533), but it seems to be a low priority question.

Anyway, it turned out that the fix is relatively easy for `vscode-neovim`. Instead of using `ctrl+[` it is possible to use `ctrl+[BracketLeft]` in keybindings definitions, which works on both OSes as expected.

I am ready to submit a PR for this change (and doing the similar change for `ctrl+]` --> `ctrl+[BracketRight]`). The only question is if I should *replace* the original definitions or *add* the new ones. The former is cleaner solution, but the latter might be more compatible?

**Reply by justinmk** [reply]
That will be very appreciated!

> question is if I should _replace_ the original definitions or _add_ the new ones

`ctrl+[` works for me on macos, so removing that seems undesirable. Adding an alias for `ctrl+[BracketRight]` seems useful.

Also in your PR, would be helpful to mention `Developer: Toggle Keyboard Shortcuts Troubleshooting` at https://github.com/vscode-neovim/vscode-neovim#troubleshooting

**Reply by risa2000** [reply]
> That will be very appreciated!

Well, I guess I might have been too ambitious and would need some time, since I realized, the bindings are generated and I would have to mod the generator :).

> > question is if I should _replace_ the original definitions or _add_ the new ones
> 
> `ctrl+[` works for me on macos, so removing that seems undesirable. Adding an alias for `ctrl+[BracketRight]` seems useful.

I guess `ctrl+[` works for macOS, as long as you use US keyboard layout or similar one, which does have `[` on the same key as English one. `ctrl+[BracketLeft]` works on the US layout the same way as `ctrl+[`. The only concern could be if there are some layouts which define `ctrl+[` not on a `[` key, but on `[` character, which is impossible to test.

From the logs I recorded (and posted on VSCode Issues link) it seems that at least on macOS and Windows, VSCode in both cases maps the keypress of the keys `Ctrl+<key right to P>` to `ctrl+[BracketLeft]` and `Ctrl+<2nd key right to P>` to `ctrl+[BracketRight]` and not to `ctrl+[`, or `ctrl+]`, which is what defines `package.json` in this extension. It just happens that on Windows VSCode somehow uses `ctrl+[` (or `ctrl+]`) as an alternative finding for that mapping while on macOS it does not.

> Also in your PR, would be helpful to mention `Developer: Toggle Keyboard Shortcuts Troubleshooting` at https://github.com/vscode-neovim/vscode-neovim#troubleshooting

You mean just mention it as a helper tool to debug keyboard issues, right?
